### PR TITLE
Make updating python/node dependencies in Docker much faster and more intuitive

### LINKED ISSUE #1230: Building and changing node/python dependencies in docker is slow
It's currently a bit of a hassle to add a single node or python dependency to our docker setup because of the fact that `docker-compose build` is running on a container that has no recollection of having installed or fetched any packages in the past--therefore it needs to re-download and re-install every single node/python package (and in the case of node, python dependencies also have to be re-downloaded/installed due to the fact that `pip install` occurs later in the `Dockerfile` than `npm install`).

This could get particularly painful once #1055 lands and we have a bunch of scientific python dependencies in addition to our existing ones; aside from updating dependencies, just *building* them could become a hassle.

Possible solutions:

1. I'm not sure if it's possible, but it'd be neat to be able to store the npm and pip caches in a Docker volume or something, so that at least re-installing the dependencies doesn't involve re-downloading them.
2. With [another open-source project I worked on](https://github.com/govlab/noi2#changing-the-dockerfile), we actually used *two* Dockerfiles: one was the "stable" base image that had most of our dependencies; it was stored on Docker Hub and made it really fast for devs to get started, since their machines could just download the Docker Hub image and not have to pip/npm install anything.  (It also made it really fast to run tests on Travis with!)  Then we had a second Dockerfile which trivially inherited from the first one and only installed the latest dependencies that were added as a result of recent PRs.  On a regular basis we'd migrate dependencies from the second Dockerfile to the first one in a way that didn't disrupt developer/deployment/CI workflows.

The second solution is one I've tried before and am pleased with, and also works nicely with Travis--thereby addressing the problem of having super slow CI builds whenever we add certain dependencies like numpy--so maybe we should just go with that. (It would obviously require changing Travis to run the *dockerized* version of our app, but doing that isn't hard.)

Also somewhat related: #569

### PR DESCRIPTION
**Pull request by toolness**
This attempts to resolve #1230 by moving the directories containing node and python dependencies out to docker volumes.

This makes updating these dependencies a *lot* more like running a local development setup: just run `pip install -r requirements-dev.txt` and `npm install` in the container and you're done.  If a single dependency changes, it's the only one that needs to be downloaded and installed.  You can even install your favorite debugging libraries this way, and the changes persist because the dependencies are in docker volumes, rather than part of the docker image itself.

It also means that there's a bit of extra setup, though: while devs likely won't have to run `docker-compose build` nearly as often, they *will* have to run `pip install` and `npm install` now.

To facilitate this, I've added an `update.sh` script that runs these, as well as `manage.py migrate` and `manage.py initgroups` all at once.  Folks developing locally (without docker) can run this, while those with Docker can run `./docker-update.sh` to get the same thing done.

I am pretty sure this should work fine with cloud-based Docker deploys, but I haven't actually tried doing it yet.

## Notes

* Due to the fact that Python doesn't encapsulate all of a project's requirements into a single directory like node does, I've made the docker setup use a virtualenv, which is the closest approximation to that. Basically, `docker_django_management.py` now detects whether the virtualenv exists--which it won't at first, because docker volumes start out empty--and if it doesn't, it creates it.  Then it transparently activates the virtualenv before running whatever command needs to be run.

* The fact that `./update.sh` runs `manage.py migrate` means that things might explode if e.g. the developer is switching between two different branches that both make migrations or something. Obviously this has always been a problem, but it might be worth it to offer some kind of advice to devs if migration fails, e.g. telling them how to easily blow away their db and start from scratch, if that's the easiest solution.

* Um, our `setup.md` could use some help. Rather than fix it now though, imma file a separate bug.

## To do

- [x] Factor out all the new CALC-specific stuff from `docker_django_management.py` somehow, so it stays nice and reusable.
- [x] We could potentially add an `update-docker.sh` script in the root dir so that docker devs can just run it instead of the more verbose `docker-compose run app ./update.sh`.  It would also allow us to run `docker-compose build` in the update script, just in case the `Dockerfile` changed.
- [x] Change `docs/docker.md` to document this new setup.

### CODE REVIEW

**Comment by toolness** [comment]
@jseppi do you have thoughts on this approach?

File: update.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+pip install -r requirements-dev.txt
+
+echo "----- Updating Node Dependencies -----"
+npm install
+
+echo "----- Migrating Database -----"
+python manage.py migrate --noinput
+
+echo "----- Initializing Groups -----"
+python manage.py initgroups
```
Comment (toolness) [inline update.sh]: This file obviously has some similarities with `cf.sh`... Not sure if it's worth it to refactor in such a way that code is reused or not.  Meh.

**Comment by jseppi** [comment]
Approach seems solid to me. I haven't tried actually running yet since I'm on a flight and don't want to stress my in-flight wifi getting containers.

**Comment by toolness** [comment]
Oh man you GOTTA get containers on a plane!  Push it to the limit man!!!!!!

**Comment by jseppi** [comment]
I've had it with this containers on this plane!

**Comment by toolness** [comment]
Ok @jseppi I think this is ready for review!!

File: docs/docker.md [hunk 2]
```diff
@@ -57,9 +55,17 @@ if it detects that Django isn't installed.
 
 ### Updating the containers
 
-All the project's dependencies, such as those mentioned in `requirements.txt`,
-are contained in Docker container images.  Whenever these dependencies change,
-you'll want to re-run `docker-compose build` to rebuild the containers.
+Whenever you update your repository via e.g. `git pull` or
+`git checkout`, you should update your containers by running
+`./docker-update.sh`.
```
Comment (jseppi) [inline docs/docker.md]: This file wasn't executable upon checkout of this branch. Is there some git magic to check it in as executable? Maybe this: http://www.pixelninja.me/make-script-committed-to-git-executable/

Comment (jseppi) [inline docs/docker.md]: Oh, after I did `chmod +x docker-update.sh`, git is showing a file difference that can be committed, so that looks easy to fix.

Comment (toolness) [inline docs/docker.md]: Ah yeah, sorry about that, I think I just forgot to `chmod +x` it before committing!

**Pull request merged**